src/main/webapp/view/discussion/discussion.js

```diff
@@ -200,20 +115,130 @@ function getNearbyDiscussionComments(timestampMs) {
+   * client needs to organize them before displaying.
+   */
+  structureComments(allComments) {
+    const commentIds = {};
+    for (const comment of allComments) {
+      comment.replies = [];
+      commentIds[comment.commentKey.id] = comment;
+    }
+
+    const rootComments = [];
+    for (const comment of allComments) {
+      if (comment.type === COMMENT_TYPE_REPLY) {
+        const parent = commentIds[comment.parentKey.value.id];
+        parent.replies.push(comment);
+      } else {
+        // Top level comments don't have parents.
+        rootComments.push(comment);
+      }
+    }
+    // Sort comments such that earliest timestamp is first.
+    rootComments.sort((a, b) => (a.timestampMs.value - b.timestampMs.value));
+    return rootComments;
+  }
+
+  /**
+   * Fetches and returns all comments in `this.#lecture` from the `ENDPOINT`.
+   * This returns a sorted array of all root comments, with replies added to
+   * their parents.
+   */
+  async fetchRootComments() {
+    const url = new URL(DiscussionManager.#ENDPOINT, window.location.origin);
+    url.searchParams.append(
+        DiscussionManager.#PARAM_LECTURE, this.#lecture.key.id);
+
+    const request = await fetch(url);
+    const json = await request.json();
+
+    return this.structureComments(json);
+  }
+
+  /**
+   * Posts `content` reloads the discussion. Adds query
+   * parameters from `params` to the request. Different types of comments
+   * require different parameters, such as `PARAM_TIMESTAMP` or `PARAM_PARENT`.
+   * The caller should ensure the correct parameters are supplied for the type
+   * of comment being posted.
+   */
+  async postComment(content, params) {
+    const url = new URL(DiscussionManager.#ENDPOINT, window.location.origin);
+    url.searchParams.append(
+        DiscussionManager.#PARAM_LECTURE, this.#lecture.key.id);
+    for (const param in params) {
+      // This is recommended by the style guide, but disallowed by linter.
+      /* eslint-disable no-prototype-builtins */
+      if (params.hasOwnProperty(param)) {
+        url.searchParams.append(param, params[param]);
+      }
+      /* eslint-enable no-prototype-builtins */
+    }
+
+    await fetch(url, {
+      method: 'POST',
+      body: content,
+    });
+  }
+
+  /**
+   * Posts `content` as a new root comment at `timestampMs` with the specified
+   * `type`.
+   */
+  async postRootComment(content, timestampMs, type) {
+    await this.postComment(content, {
+      [DiscussionManager.#PARAM_TIMESTAMP]: timestampMs,
+      [DiscussionManager.#PARAM_TYPE]: type,
+    });
+  }
+
+  /**
+   * Posts `content` as a reply to `parentId`.
+   */
+  async postReply(content, parentId) {
+    await this.postComment(content, {
+      [DiscussionManager.#PARAM_PARENT]: parentId,
+      [DiscussionManager.#PARAM_TYPE]: COMMENT_TYPE_REPLY,
+    });
+  }
+}
+
+
 /**
  * Renders a comment and its replies, with a form to post a new reply.
  */
 class DiscussionComment extends HTMLElement {
```

Comment:
Same here, move to separate file

Reply:
Going to wait on moving this one for a future PR.

Reply:
Please add a TODO for this, in general if there is anything that will be done in the future for discussion.js a TODO will be most helpful

Reply:
Added